gdpr/purposes/default.py

```diff
@@ -1,31 +1,136 @@
-from datetime import timedelta
-
-from collections import OrderedDict
+from collections import namedtuple
+from typing import Any, Dict, Iterable, KeysView, List, Optional, TYPE_CHECKING, Tuple, Type, Union
 
 from django.core.exceptions import ImproperlyConfigured
+from django.db.models import Model, Q
+
+from gdpr.loading import anonymizer_register, purpose_register
+
+if TYPE_CHECKING:
+    from gdpr.models import LegalReason
+    from gdpr.anonymizers import ModelAnonymizer
 
+FieldList = Union[List[str], Tuple, KeysView[str]]  # List, tuple or return of dict keys() method.
+FieldMatrix = Union[str, Tuple[Any, ...]]
+RelatedMatrix = Dict[str, FieldMatrix]
+PurposesFieldMatrices = Iterable[FieldMatrix]
+PurposesFieldLists = Iterable[FieldList]
+PurposesRelatedFieldMatrices = Iterable[RelatedMatrix]
+PurposesSplitFields = Iterable["PurposeSplitFields"]
 
-purposes_map = OrderedDict()
+
+class PurposeSplitFields(namedtuple('PurposeSplitFields', ['local', 'related'])):
+    def __new__(_cls, local=tuple(), related=dict()):
+        return super().__new__(_cls, local, related)
+
+    def get_tuple(self):
+        if type(self.local) == str:
+            return (self.local, *tuple(self.related.items()))
+        return (*self.local, *tuple(self.related.items()))
 
 
 class PurposeMetaclass(type):
 
     def __new__(mcs, name, bases, attrs):
+        from gdpr.loading import purpose_register
+
         new_class = super(PurposeMetaclass, mcs).__new__(mcs, name, bases, attrs)
         if hasattr(new_class, 'slug') and new_class.slug:
-            if new_class.slug in purposes_map:
+            if new_class.slug in purpose_register:
                 raise ImproperlyConfigured('More anonymization purposes with slug {}'.format(new_class.slug))
 
-            purposes_map[new_class.slug] = new_class
+            purpose_register.register(new_class.slug, new_class)
         return new_class
 
     def __str__(self):
         return str(self.name)
 
 
 class AbstractPurpose(metaclass=PurposeMetaclass):
+    """
```

Comment:
slušně složitá třída jen kvůli tomu, že je potřeba vytáhnout správná pole. Moc se mi nelíbí řešení přes PurposeSplitFields. Je to takové dost nečitelné